src/model_theory/basic.lean

```diff
@@ -366,5 +388,228 @@ lemma comp_assoc (f : M ≃[L] N) (g : N ≃[L] P) (h : P ≃[L] Q) :
+{ on_function := λ n, sum.map (λ f, ϕ.on_function f) (λ f, ψ.on_function f),
+  on_relation := λ n, sum.map (λ f, ϕ.on_relation f) (λ f, ψ.on_relation f) }
+
+instance sum_elim_is_expansion_on {L'' : language} (ψ : L'' →ᴸ L') (M : Type*)
+  [L.Structure M] [L'.Structure M] [L''.Structure M]
+  [ϕ.is_expansion_on M] [ψ.is_expansion_on M] :
+  (ϕ.sum_elim ψ).is_expansion_on M :=
+⟨λ _ f _, sum.cases_on f (by simp) (by simp), λ _ R _, sum.cases_on R (by simp) (by simp)⟩
+
+instance sum_map_is_expansion_on {L₁ L₂ : language} (ψ : L₁ →ᴸ L₂) (M : Type*)
+  [L.Structure M] [L'.Structure M] [L₁.Structure M] [L₂.Structure M]
+  [ϕ.is_expansion_on M] [ψ.is_expansion_on M] :
+  (ϕ.sum_map ψ).is_expansion_on M :=
+⟨λ _ f _, sum.cases_on f (by simp) (by simp), λ _ R _, sum.cases_on R (by simp) (by simp)⟩
+
+end Lhom
+
+section constants_on
+variables (α : Type u')
+
+/-- A language with constants indexed by a type. -/
+def constants_on : language.{u' 0} := ⟨λ n, nat.cases_on n α (λ _, pempty), λ _, pempty⟩
+
+variables {α}
+
+@[simp] lemma constants_on_const : (constants_on α).const = α := rfl
+
+instance is_algebraic_constants_on : is_algebraic (constants_on α) :=
+language.is_algebraic_of_empty_relations
+
+instance is_relational_constants_on [ie : is_empty α] : is_relational (constants_on α) :=
+⟨λ n, nat.cases_on n ie (λ _, pempty.is_empty)⟩
+
+/-- Gives a `constants_on α` structure to a type by assigning each constant a value. -/
+def constants_on.Structure (f : α → M) : (constants_on α).Structure M :=
+{ fun_map := λ n, nat.cases_on n (λ a _, f a) (λ _, pempty.elim),
+  rel_map := λ _, pempty.elim }
+
+variables {β : Type v'}
+
+/-- A map between index types induces a map between constant languages. -/
+def Lhom.constants_on_map (f : α → β) : (constants_on α) →ᴸ (constants_on β) :=
+⟨λ n, nat.cases_on n f (λ _, pempty.elim), λ n, pempty.elim⟩
+
+end constants_on
+
+section with_constants
+
+variable (L)
+
+section
+variables (α : Type w)
+
+/-- Extends a language with a constant for each element of a parameter set in `M`. -/
+def with_constants : language.{(max u w) v} := L.sum (constants_on α)
+
+localized "notation L`[[`:95 α`]]`:90 := L.with_constants α" in first_order
+
+/-- The language map adding constants.  -/
+def Lhom_with_constants : L →ᴸ L[[α]] := Lhom.sum_inl
+
+variable {L}
+
+/-- Adds constants to a language map.  -/
+def Lhom.add_constants {L' : language} (φ : L →ᴸ L') :
+  L[[α]] →ᴸ L'[[α]] := φ.sum_map (Lhom.id _)
+
+instance params_Structure (A : set α) : (constants_on A).Structure α := constants_on.Structure coe
+
+variables (L) (α)
+
+/-- The language map removing an empty constant set.  -/
+def Lhom_trim_empty_constants [is_empty α] : L[[α]] →ᴸ L :=
+Lhom.sum_elim (Lhom.id L) (Lhom.of_is_empty (constants_on α) L)
+
+variables {α} {A B : set α} (h : A ⊆ B)
+
+/-- The language map extending the constant set.  -/
+def Lhom_constants_inclusion : L[[A]] →ᴸ L[[B]] :=
+Lhom.sum_map (Lhom.id L) (Lhom.constants_on_map (set.inclusion h))
```

Comment:
I think this shouldn't exist, and you just want the function `(f : α → β) : L[[α]] → L[[β]]`